### ISSUE TITLE
route.params is undefined, nicer if it was an empty Object {} by default

### ISSUE THREAD

**Issue by Jarred-Sumner** [react-navigation/react-navigation#6851]
React Navigation v5's API is so much better! This is a tiny thing.

**Current Behavior**
When there are no `params` passed to `navigator.navigate`, `route.params` is undefined.  Additionally, the type definition doesn't indicate that `params` can be `undefined`.

This means that the following code will raise an exception because `params` is not an `Object`:
```javascript
const {showStart, onChange} = useRoute().params
```
![image](https://user-images.githubusercontent.com/709451/74502593-d1d09e00-4ea2-11ea-88a5-cb1daf889c14.png)

This makes sense. There are no params, so `params` is never set.

**Expected Behavior**

From a developer experience perspective though, It'd be a little easier if it was an empty `Object` when undefined. That way, I can use destructuring without optional chaining.

**How to reproduce**

Navigate to a route but don't pass any params to it.

```javascript
const {showStart, onChange} = useRoute().params
```

**Your Environment**
```
├─ @react-navigation/core@5.1.2
│  ├─ @react-navigation/routers@^5.0.1
├─ @react-navigation/native-stack@5.0.3
├─ @react-navigation/native@5.0.3
│  └─ @react-navigation/core@^5.1.2
├─ @react-navigation/routers@5.0.1
├─ @react-navigation/stack@5.0.3
├─ react-navigation-animated-switch@0.4.1
├─ react-navigation-shared-element@0.5.0
```

--

In terms of how to implement this...maybe it could be an `Object.defineProperty` getter on the `route` object returned by `useRoute`?

**Comment by Grohden** [comment]
I'm curious:

> Additionally, the type definition doesn't indicate that params can be undefined.

You mean that every TS dev should deal with `undefined` even when they know that it will never be undefined?

I would rather use this from [the guides](https://reactnavigation.org/docs/typescript 
):
```
type RootStackParamList = {
  Home: undefined;
  Profile: { userId: string };
  Feed: { sort: 'latest' | 'top' } | undefined;
};
```
And annotate the provided hooks properly (https://reactnavigation.org/docs/typescript#annotating-useroute)


If you type things correctly, you **probably** will not run into these kind of issues, alternatively you could use:
```
const { showStart, onChange } = useRoute().params ?? {}
```

**Comment by andreasmischke** [comment]
> You mean that every TS dev should deal with `undefined` even when they know that it will never be undefined?

That's the current state and should *not* be the case.

> If you type things correctly, you **probably** will not run into these kind of issues

Unfortunately that's not the case. By typing things correctly, you can define that all of your `params` are optional. But that should not lead to the `params` object itself being undefined. Even if it does (which might be okay) the TS annotations should tell you that `params` is `something™️ | undefined`.

To say things simple: The TS annotations currently **do not** match the real behaviour of the library.

**Comment by Grohden** [comment]
> That's the current state and should not be the case.

I still don't get it, you don't need to deal with `undefined` with current state, as I said, if you type things correctly you will not get a `undefined` as a param unless you type it as `undefined` and/or use ''unsafe'' calls. (there's a fault at the library typings: you can call `navigate(string)` and ts will not complain even when the route is required to have params)

> By typing things correctly, you can define that all of your params are optional. But that should not lead to the params object itself being undefined. 

Params will not be `undefined` if you use the types properly:
```typescript
type RoutesParamList = {
   Foo: Partial<{ foo: string }>
   Bar: Partial<{ bar: number }> | undefined
}
```
`Foo` will need to be called with at least an empty object (`navigate('Foo', {})`), `undefined` is not accepted here.
`Bar` will accept `undefined` (`navigate('Bar', undefined)`)

> To say things simple: The TS annotations currently do not match the real behavior of the library.

Yeah... you have a point about the behavior, there's always the chance that the lib will send a `undefined` because it can send it by default. 
But by adding `| undefined` at the lib, everyone will need to deal with `undefined` even when they make sure every call to navigate is using 'safe' types/correct generics, and by adding `| {}` at the lib, everyone will need to deal with optional arguments always.

Letting you choose how you deal with your types moves the responsibility to you, you chose what you deal with.

Also if you use `{}` instead of `undefined` (not adding the types and strict type check), you gonna have to deal with `undefined` in the end:
 * Using `undefined` you get an error accessing any prop in it.
 * Using `{}` you get `undefined` accessing props, its easier to pass them down and see errors somewhere else.

**For me** the last option is bad.

**Comment by Shywim** [comment]
As said in the issue's description, the expected behavior is to have an *empty* object by default. There's no reason to have to deal with `undefined` on `params` and `params` (or equivalent) is never `undefined` in other libs like react-router or just in the browser's native Location api.

**Comment by Grohden** [comment]
> whereas there is no bad implication I know of to have an empty params object when there's no param

Here's one TS specific [issue](https://github.com/microsoft/TypeScript/issues/12815#issuecomment-266250230), [here's the playground](https://www.typescriptlang.org/play/?jsx=3#code/C4TwDgpgBAQghgJygXigbygI0QLigZ2AQEsA7AcygF8AoG0SKAJQHsBXYCBAHgBUA+FOhpRRUMIjgBbfHl5QAPulq0aAMzakAxsGItSUNvgisOXPvwAUASjynOPAcLFQEEYGwQG0Il2IkI0rLKvqIqdBrauvpQAMIsUmD6EKTAAII2zmJa+oTo4pIy1EJGJuwO3PAIVtahUAD09VAA8gAWUKQsADRQwPhQACYsEPikAOTAUADWnQDuUHCY5ViIUFpsAF51jVDEk8CtZFP9B3CTe1BScCBY0GhUC6QDUCDsa3DjwNtNcFpaI-1MBAtHBSr1WlwIGN+p0oGNsAgxtMIDcyLsvi4cqQ8hgEUIxgMIGpQQAbYBIh6oAJBOguNweLwrBA0VSRHR6AzxRLJVIwTI+TG5SYYalFSmGYz2cxVGq0sQ7AByLEgPWALBYlzYWnaYzcahJwOi4ygahYSAY0HIbEQA3wdWIaiglmamAAVoaAHRTFH4Syi-DWD0GigHKCCAAM1iyflc7k8Bn9HoRdVUdLjjNIbBJJJZdHUmnZMS5SVIKWAsX5dSxOIKgTFJUl5WliFl3ygSvmfWmc36-t2-Sq9sdlnhiCRaP9UYFMdjDIThXwScQKblonp8Y6WZztCAA)

But maybe this ts issue could be solved by the router receiving a default value... :thinking: 

```typescript
const useRouterWithDefs = <R, T>(default: T) => {
   const router = useRouter<R>()

   return useMemo(() => ({
       ...router,
       params: router.params === undefined ? default : router.params
   }), [router])
}
```
Anyway, as I said, _for me_ is not good, but it's not so bad either, I use types correctly and doesn't receive `undefined` at my routes, so it doesn't make much a difference for me if it does not force me to validate undefined and/or empty objects

**Comment by vinipachecov** [comment]
Excuse me @Grohden but I could not follow the guides: Why does the second parameter of the RouteProp has be inside params list?

In the types.d.ts it was meant to be the actual RouteName, wasn't it? Dunno but looks confusing to me,.

Why this should be invalid ?

```ts

type RootStackParamList = {
  vaccine: string
};

type RouteScreenProp = RouteProp<Partial<{
  vaccine: VaccineModel;
}>, 'Profile'>;
```

All I want to type is that a specific route param will have a vaccine prop. I don't want to name the route 'vaccine'. Did I miss something?

**Comment by andreasmischke** [comment]
@vinipachecov Your `RootStackParamList` should contain the parameters for all your routes:

```typescript
type RootStackParamList = {
  Profile: {
    vaccine: string // Or VaccineModel? You have to decide on one
  }
};

type RouteScreenProp = RouteProp<RootStackParamList, 'Profile'>;
```

**Comment by Grohden** [comment]
@vinipachecov wut?
The proper usage should be:
```typescript

type RootStackParamList = {
  // Your route is named 'Profile', it has this object as params:
  Profile: Partial<{
      vaccine: VaccineModel
  }>
};

// here you say to ts 'take the 'Profile' entry type from RootStackParamList and use it as a type for navigation
type RouteScreenProp = RouteProp<RootStackParamList, 'Profile'>;
```

**Comment by vinipachecov** [comment]
Thank you @Grohden! I wasn't understanding the purpose of using the name to refer to a prop inside the param list.

**Comment by martintreurnicht** [comment]
I've run into this issue too, it's definitely and issue with the types work with regards to navigation params, consider the following route param list:

```typescript
type RootStackParamList = {
  Profile: {
    vaccine: string
  }
  Home: undefined
};
```

If I try and do the following:
```typescript
navigation.navigation('Profile')
```
I would expect an error, because params can't be undefined

But, will work for Home, because params type is undefined

Maybe this is a limitation to typescript's type system. The easiest would be to have params just be an object always, and if no params is specified, it's an empty object. Either that or the type for route.params not be made optional

**Comment by satya164** [comment]
@martintreurnicht that seems like a bug. can you open a bug report about it?

**Comment by Grohden** [comment]
> (there's a fault at the library typings: you can call navigate(string) and ts will not complain even when the route is required to have params)
@satya164 @martintreurnicht I think that this could fix the issue:

```typescript
type RootStackParamList = {
    Profile: {
        vaccine: string
    }
    Home: undefined
};

export type PropsOfType<T, O> = ({
    [P in keyof T]: T[P] extends O ? P : never
})[keyof T];

type Nullables = PropsOfType<RootStackParamList, undefined>
type NonNullables = keyof Omit<RootStackParamList, Nullables>

interface NavigationInterface {
    navigate(route: Nullables): void
    navigate<key extends NonNullables>(
        route: key,
        args: RootStackParamList[key]
    ): void
}

declare const navigation: NavigationInterface
'
navigation.navigate('Home')
navigation.navigate('Home', undefined) // Nope, cant explicity pass undefined.
navigation.navigate('Profile') // Not a nullable route'
```

But TS is very bad at showing you error messages :/ the above code doesn't make clear to you that you should be 'using' the second function definition

**Comment by satya164** [comment]
@Grohden can you send a PR?

**Comment by Grohden** [comment]
@satya164 not right now.. probably next week I'll open

**Comment by martintreurnicht** [comment]
> @martintreurnicht that seems like a bug. can you open a bug report about it?

I don't think it's a bug as much as it's a limitation to how the type system works right now. I will attempt at creating a PR when i have some free time

**Comment by swayok** [comment]
I've just upgraded form v4 to v5 and now having problems with this `this.props.route.props is undefined` errors which were totally not expected as default behavior. 
For example I have a screen with route props like this:
```
export type AppStackParamList = {
    Apartments: {
        goBackAfterChange?: boolean
    }
}
``` 
This way inside screen class I expect `this.props.route.params` to be Object with optional property `goBackAfterChange` so that I can navigate like this: `navigator.navigate('Apartments')` and in Apartemnts screen do something like this: `if (this.props.route.params.goBackAfterChange) {}`. But instead I get error `this.props.route.params is undefined`. It is totally illogical behavior for most developers. For what reason and use case `this.props.route.params` by default should be undefined? I was trying to think of a situation where such behavior would be useful but there seems to be none. When you do not expect params - you do not check `this.props.route.params` to be undefined...  On other side you never expect `this.props.route.params` to be undefined instead of an object. 

It is quite strange to force developers do this: `navigator.navigate('Apartments', {})` or this: `if (this.props.route.params && this.props.route.params.goBackAfterChange) {}` and never forget about such behavior.

**Comment by wuhuanda** [comment]
Same idea. I think the route.params should be an {} by default.

**Comment by micnguyen** [comment]
Yeah, I would also like this. For those in JS land and still would like to destructure, you can do this:
```
const { foo, bar } = route.params ?? { foo: null, bar: null }
```
Which will use the values from `route.params` if it exists, otherwise will become `null` and not crash from accessing it. This approach you can still use `navigation.navigate(aRoute)` without a second param `{}` prop.

**Comment by wuhuanda** [comment]
> Yeah, I would also like this. For those in JS land and still would like to destructure, you can do this:
> 
> ```
> const { foo, bar } = route.params ?? { foo: null, bar: null }
> ```
> 
> Which will use the values from `route.params` if it exists, otherwise will become `null` and not crash from accessing it. This approach you can still use `navigation.navigate(aRoute)` without a second param `{}` prop.

Usually, I do like this:
```
const {foo, bar} = route.params || {}
```
But, I think it is ugly. Every place I use the route, I must write the `||{}`.

**Comment by danyalasif** [comment]
> Yeah, I would also like this. For those in JS land and still would like to destructure, you can do this:
> 
> ```
> const { foo, bar } = route.params ?? { foo: null, bar: null }
> ```
> 
> Which will use the values from `route.params` if it exists, otherwise will become `null` and not crash from accessing it. This approach you can still use `navigation.navigate(aRoute)` without a second param `{}` prop.

You can also do this for a cleaner look 

```const { foo = null, bar = null } = route.params ?? { }```

**Comment by euroclydon37** [comment]
Plus one for empty object default value.

**Comment by elhoucinOlmstaf** [comment]
I have the same issue

Object { key: "bbb-EhG7Ouzw79ZQ8NNIcVBb_", name: "bbb", params: undefined, … }
key: "bbb-EhG7Ouzw79ZQ8NNIcVBb_"
name: "bbb"
params: undefined

it works just fine with stackNavigation , but when I add DrawerNavigator . it does not work any more. 
any solutions please!!

**Comment by IyiolaOwabumowa** [comment]
Sigh

**Comment by adrenaline681** [comment]
I agree, route.params should just be an empty object if no parameters have been passed

**Comment by iMarvinS** [comment]
Any progress on this one?

**Comment by tobzy** [comment]
Any progress on this one? Was it fixed in newer versions?

**Comment by nica0012** [comment]
Any news here?I migrated over to React Navigation and dealing with undefined from route.param values is not fun...

**Comment by jaexplorer** [comment]
Any update?

**Comment by Bastianowicz** [comment]
Isn't it just [this line](https://github.com/react-navigation/react-navigation/blob/aa5ae177d4d6b7ea2ae36c194b6a46747c242f9b/packages/core/src/getStateFromPath.tsx#L596) that needs a little update?
Is there any collateral damage to be expected?

**Comment by intomyam** [comment]
I fixed only the type for now. https://github.com/react-navigation/react-navigation/pull/11795

**Comment by intomyam** [comment]
@satya164 
You closed https://github.com/react-navigation/react-navigation/pull/11795, but the underlying problem is that the `params`  can be undefined (even if I define that `params` is not undefined when using useRoute).
This is a behavior that could crash my app unexpectedly, I have adapted the type of useRoute to the actual value as a temporary fix.
I suppose changing `params` to Object instead of undefined would be a better approach.
